### ISSUE TITLE
Build Tooling: Optimize build by caching results

### ISSUE THREAD

**Issue by aduth** [WordPress/gutenberg#15967]
Previously: #15230 , #13124 

As explored previously in #13124 and discussed in #15230, we may be able to benefit from improved build times by leveraging a persisted cache of previously-built artifacts.

**Implementation Notes:**

- Before running a JavaScript or SASS file's build tasks, consider whether a cached result exists
   - Cache can exist in some `gitignore`d location, e.g. `node_modules/.cache/gutenberg-build` or `require( 'os' ).tmpdir()` (see [`babel-loader`](https://github.com/babel/babel-loader) as prior art)
   - Files are named by hash of the source file's content, to assure reuse only on identical source
      - Use a fast hash like SHA-1 or MD5
      - See: https://github.com/WordPress/gutenberg/pull/15230#discussion_r288550607
- Use [Travis caching](https://docs.travis-ci.com/user/caching/) to reuse artifacts between builds
   - This means the cache location must be configurable, e.g. similar to `BABEL_CACHE_DIRECTORY` override in #14860
   - To keep Travis cache download time reasonable, we must prune cache, ideally automatically
      - See: https://github.com/WordPress/gutenberg/pull/15230#issuecomment-497565963, https://github.com/WordPress/gutenberg/pull/15230#issuecomment-498289949

**Comment by aduth** [comment]
A complication here is that one source file will have multiple output files, so it's not quite straight-forward to detect a cached result by the presence of a hash.

A couple ideas:

- We could refactor the build worker to have a function which produces the set of filenames which would result from a build, which is used both for checking for a cached result, and for the build itself.
- The cache lookup result could be a _directory of files_ rather than a single file

**Comment by noisysocks** [comment]
Last week I started experimenting with an approach where we cache _objects_ instead of _files_ by serialising the object as a JSON string. You can see where I got over in the [`try/build-caching` branch](https://github.com/WordPress/gutenberg/compare/try/build-caching).

It's not finished and there are errors when you run the build. Still, if anyone would like to pick this up: go for it! 

> To keep Travis cache download time reasonable, we must prune cache, ideally automatically

Thinking that we just recursively traverse the cache directory and delete files that are more than 3 months old. If traversal itself is too slow then we could only perform this operation when `Math.random() < 0.1`.

**Comment by aduth** [comment]
> Last week I started experimenting with an approach where we cache _objects_ instead of _files_ by serialising the object as a JSON string. You can see where I got over in the [`try/build-caching` branch](https://github.com/WordPress/gutenberg/compare/try/build-caching).

As written, are you assuming that we'd have multiple identical source files which could benefit from the memoization?

I could also imagine just writing this as a file with the stringified JSON object, as an alternative to what I considered in my previous comment of a directory of files, where the keys in the JSON object are equivalent to files in a directory. Unclear whether this would have a performance difference, e.g. if a file system would be faster at copying a file than to parse JSON and write a new one from the object contents.

>Thinking that we just recursively traverse the cache directory and delete files that are more than 3 months old. If traversal itself is too slow then we could only perform this operation when Math.random() < 0.1.

This makes sense, but I'm not sure _when_ this would occur. For me, ideally it'd be some "fire and forget" (non-blocking) action which occurs after a build completes.

There's various lifecycle steps like `after_success`, `after_script` ([documentation](https://docs.travis-ci.com/user/job-lifecycle/#the-job-lifecycle)).

If it's not possible to use these as "non-blocking" (i.e. assuming the container shuts down after the build completes), maybe instead it's just a separate task of the build to do the cleanup, since presumably it wouldn't take long and could be done in parallel to other tasks.

**Comment by noisysocks** [comment]
> As written, are you assuming that we'd have multiple identical source files which could benefit from the memoization?

This isn't an explicit objective of what I wrote, no.

We could add the path of the input file to the object that is hashed so that two identical source files using the same cache entry. I figured this isn't necessary though since, if two source files are identical, they will compile to the same result.

**Comment by noisysocks** [comment]
> Unclear whether this would have a performance difference, e.g. if a file system would be faster at copying a file than to parse JSON and write a new one from the object contents.

Yeah, you're right that it's likely that a `mv` is faster. 

Perhaps something like this for an e.g. `.cache/manifest.json`:

```json
{
	"src/example.js": {
		"hash": "1a79a4d60de6718e8e5b326e338ae533",
		"output": {
			"build/example.js": ".cache/1a79a4d60de6718e8e5b326e338ae533.js",
			"build/example.js.map": ".cache/1a79a4d60de6718e8e5b326e338ae533.js.map"
		}
	}
}
```

**Comment by aduth** [comment]
I spent some time experimenting with this on Friday, and make some decent (but incomplete/buggy) progress. The work is currently reflected in the [`add/build-cache` branch](https://github.com/WordPress/gutenberg/commits/add/build-cache) (specifically d2cf89e2385016937079f04ee8a6411051cac424).

A few of my observations:

- While it was natural to try to add caching as an enhancement of the current behavior in the `build-worker.js` file (the forked build process), and while it did mostly work and was fairly simple to implement, I quickly ran into an issue of race conditions. When multiple of these forked processes would try to copy over a cached result, they'd often simultaneously try creating the recursive folder structure at the same time, resulting in errors (`EEXISTS` folder exists). It might be possible to swallow these errors, but in the tools I was trying to leverage (`fs-extra`, `recursive-copy`) it was not handled gracefully.
- This led me instead to consider caching as part of a stream workflow of the managing parent process. This feels like a very appropriate use of streams, where the build was just one step in a pipeline of writing to/copying from cache.
- Cache invalidation rears its head, as always. The implementation in d2cf89e2385016937079f04ee8a6411051cac424 maintains a list of files which are relevant to the build process, whose contents are incorporated into the cache checksum hash. I fear this is quite fragile and could easily fall out of sync (resulting in wrongful cache reuse).
- The implementation of the cache is as the second point mentioned in https://github.com/WordPress/gutenberg/issues/15967#issuecomment-499078509 , where the result is a mapping of files which are written to a directory with name of a checksum in `node_modules/.cache/gutenbuild`.
   - In reflection, I think there's a minor flaw to this which makes it non-optimal, which is that in copying files from cache, each copy must occur with a recursive folder check. For example, considering `edit.js` and `save.scss` from `block-library/src/paragraph`, each would have to ensure that the destination directory (`build/paragraph`) exist or are created before copying the file. Ideally this is at least memoized to know that a prior copy for the same directory had guaranteed its existence.
- We now have `fs-extra` as a dependency (from the recent release tool introduced in #15699), which helps as a minor refactor to avoid promisification of `fs` functions (see `build-worker.js` changes in d2cf89e2385016937079f04ee8a6411051cac424)
- The necessary refactor to change the build tasks to return an object of file mappings (file name to file contents) proved beneficial regardless of the introduction of caching. In retrospect, having the build task manage the creation of its output directory added some unnecessary complexity.

**Comment by youknowriad** [comment]
I think there's probably still some aggressive caching we should be able to do today but I believe we kind of solved most of the issues already with the new build tool. The remaining bottleneck today is mostly TS. so I'm closing this issue for now.
